### PR TITLE
CalDAV calendar public sharing

### PR DESCRIPTION
**Pull request by LukasReschke**
From https://github.com/owncloud/core/pull/25351/

Needs proper review and testing as there were some smaller conflicts. Will also myself go through it again as well.

### CODE REVIEW

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 1]
```diff
@@ -117,19 +120,24 @@ class CalDavBackend extends AbstractBackend implements SyncSupport, Subscription
...
+	/** @var IConfig */
+	private $config;
 
 	/**
 	 * CalDavBackend constructor.
 	 *
 	 * @param IDBConnection $db
 	 * @param Principal $principalBackend
 	 * @param IUserManager $userManager
+         * @param IConfig $config
```
Comment (LukasReschke) [inline apps/dav/lib/CalDAV/CalDavBackend.php]: Indentation

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 2]
```diff
@@ -1443,6 +1566,46 @@ public function getShares($resourceId) {
...
+	public function setPublishStatus($value, $calendar) {
+		$query = $this->db->getQueryBuilder();
+		if ($value) {
+			$query->insert('dav_shares')
+				->values([
+					'principaluri' => $query->createNamedParameter($calendar->getPrincipalURI()),
+					'type' => $query->createNamedParameter('calendar'),
+					'access' => $query->createNamedParameter(self::ACCESS_PUBLIC),
+					'resourceid' => $query->createNamedParameter($calendar->getResourceId()),
+					'publicuri' => $query->createNamedParameter(md5($this->config->getSystemValue('secret', '') . $calendar->getResourceId()))
```
Comment (LukasReschke) [inline apps/dav/lib/CalDAV/CalDavBackend.php]: This seems horrible 💣

Comment (LukasReschke) [inline apps/dav/lib/CalDAV/CalDavBackend.php]: => Need to use random data here

File: apps/dav/lib/CalDAV/Publishing/PublishPlugin.php [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+	 *
+	 * @return null|bool
+	 */
+	public function httpPost(RequestInterface $request, ResponseInterface $response) {
+		$path = $request->getPath();
+
+		// Only handling xml
+		$contentType = $request->getHeader('Content-Type');
+		if (strpos($contentType, 'application/xml') === false && strpos($contentType, 'text/xml') === false) {
+			return;
```
Comment (LukasReschke) [inline apps/dav/lib/CalDAV/Publishing/PublishPlugin.php]: Return is void then and not null

File: apps/dav/lib/DAV/PublicAuth.php [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+use Sabre\HTTP\RequestInterface;
+use Sabre\HTTP\ResponseInterface;
+
+class PublicAuth implements BackendInterface {
+
+	/** @var string[] */
+	private $publicURLs;
+
+	/**
+	 * @param string[] $publicURLs
```
Comment (LukasReschke) [inline apps/dav/lib/DAV/PublicAuth.php]: The cake is a lie.

File: apps/dav/lib/DAV/PublicAuth.php [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+			return [true, "principals/system/public"];
+		}
+		return [false, "No public access to this resource."];
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	function challenge(RequestInterface $request, ResponseInterface $response) {
+	}
```
Comment (LukasReschke) [inline apps/dav/lib/DAV/PublicAuth.php]: Seems odd to me.

File: apps/dav/lib/DAV/PublicAuth.php [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+	/**
+	 * @param RequestInterface $request
+	 * @return bool
+	 */
+	private function isRequestPublic(RequestInterface $request) {
+		$params = $request->getQueryParameters();
+		if (isset($params['sabreAction']) && $params['sabreAction'] == 'asset') {
```
Comment (LukasReschke) [inline apps/dav/lib/DAV/PublicAuth.php]: Needs a check if that could be bypassed somehow

Comment (LukasReschke) [inline apps/dav/lib/DAV/PublicAuth.php]: Yes. We can't implement it that way. This is dangerous…

File: apps/dav/lib/CalDAV/Publishing/PublishPlugin.php [hunk 7]
```diff
@@ -0,0 +1,215 @@
+<?php
+
+namespace OCA\DAV\CalDAV\Publishing;
```
Comment (nickvergessen) [inline apps/dav/lib/CalDAV/Publishing/PublishPlugin.php]: missing copyright

File: apps/dav/lib/CalDAV/Publishing/Xml/Publisher.php [hunk 8]
```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace OCA\DAV\CalDAV\Publishing\Xml;
```
Comment (nickvergessen) [inline apps/dav/lib/CalDAV/Publishing/Xml/Publisher.php]: missing header

**Comment by tcitworld** [comment]
Please review again. :)

**Comment by MorrisJobke** [comment]
@schiessle Signed-off is missing here - how to handle this case?

**Comment by LukasReschke** [comment]
> @schiessle Signed-off is missing here - how to handle this case?

Ignore. I won't touch every old PR and add a "Signed-off" at every commit. Otherwise, @schiessle can happily volunteer for doing so 🚀

**Comment by tcitworld** [comment]
I can squash mine for what it's worth, but I guess you don't want all of them squashed.

**Comment by LukasReschke** [comment]
Let's get this in 👍

**Comment by georgehrke** [comment]
Thanks everyone who helped and thanks a lot to @tcitworld :)

**Comment by vasyugan** [comment]
Is this in NC 11?

**Comment by nickvergessen** [comment]
yes

**Comment by georgehrke** [comment]
Yes, it is

**Comment by tcitworld** [comment]
You guys are too fast for me.

**Comment by LukasReschke** [comment]
Yes

**Comment by blizzz** [comment]
да

**Comment by jospoortvliet** [comment]
No!

Wait, what? 

👍

**Pull request merged**